test/hotspot/jtreg/compiler/lib/template_framework/CodeFrame.java

```diff
@@ -0,0 +1,156 @@
+ * but we jump to an outer {@link CodeFrame}, ensuring that we insert {@link Code} at the outer frame,
+ * and operating on the names of the outer frame. Once the {@link Hook#insert}ion is complete, we jump
+ * back to the caller {@link TemplateFrame} and {@link CodeFrame}.
+ */
+class CodeFrame {
+    public final CodeFrame parent;
+    private final List<Code> codeList = new ArrayList<Code>();
+    private final Map<Hook, Code.CodeList> hookCodeLists = new HashMap<>();
+
+    /**
+     * The {@link NameSet} is used for variable and fields etc.
+     */
+    final NameSet names;
+
+    private CodeFrame(CodeFrame parent, boolean isTransparentForNames) {
+        this.parent = parent;
+        if (parent == null) {
+            // NameSet without any parent.
+            this.names = new NameSet(null);
+        } else if (isTransparentForNames) {
+            // We use the same NameSet as the parent - makes it transparent.
+            this.names     = parent.names;
+        } else {
+            // New NameSet, to make sure we have a nested scope for the names.
+            this.names     = new NameSet(parent.names);
+        }
+    }
+
+    /**
+     * Creates a base frame, which has no {@link #parent}.
+     */
+    public static CodeFrame makeBase() {
+        return new CodeFrame(null, false);
+    }
+
+    /**
+     * Creates a normal frame, which has a {@link #parent} and which defines an inner
+     * {@link NameSet}, for the names that are generated inside this frame. Once this
+     * frame is exited, the name from inside this frame are not available anymore.
+     */
+    public static CodeFrame make(CodeFrame parent) {
+        return new CodeFrame(parent, false);
+    }
+
+    /**
+     * Creates a special frame, which has a {@link #parent} but uses the {@link NameSet}
+     * from the parent frame, allowing {@link Template#defineName} to persist in the outer
+     * frame when the current frame is exited. This is necessary for {@link Hook#insert},
+     * where we would possibly want to make field or variable definitions during the insertion
+     * that are not just local to the insertion but affect the {@link CodeFrame} that we
+     * {@link Hook#set} earlier and are now {@link Hook#insert}ing into.
+     */
+    public static CodeFrame makeTransparentForNames(CodeFrame parent) {
+        return new CodeFrame(parent, true);
+    }
+
+    void addString(String s) {
+        codeList.add(new Code.Token(s));
+    }
+
+    void addCode(Code code) {
+        codeList.add(code);
+    }
+
+    void addHook(Hook hook) {
+        if (hasHook(hook)) {
+            // This should never happen, as we add a dedicated CodeFrame for each hook.
+            throw new RuntimeException("Internal error: Duplicate Hook in CodeFrame: " + hook.name());
+        }
+        hookCodeLists.put(hook, new Code.CodeList(new ArrayList<Code>()));
+    }
+
+    boolean hasHook(Hook hook) {
```

Comment:
Can be made private:
```suggestion
    private boolean hasHook(Hook hook) {
```